File: electricity-bill-calculator/app/schemas/portal.py
"""Schemas for the restricted submitter self-service API.

These models intentionally omit the legacy charge and tax columns retained on
the billing tables for historic invoice compatibility.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_serializer

# Use the same slab model returned by the tariff settings service.  The
# calculation schemas contain a structurally similar model, but Pydantic does
# not treat instances of the two classes as interchangeable.
from app.schemas.settings import ElectricitySlab
from app.schemas.electricity import BreakdownItem


class PortalModel(BaseModel):
    @field_serializer("*", when_used="json")
    def serialize_decimal(self, value):
        if isinstance(value, Decimal):
            return float(value)
        return value


class MyMeterReading(PortalModel):
    id: int
    meter_name: str
    previous_reading: Decimal
    current_reading: Decimal
    units_used: Decimal
    reading_date: date
    created_at: datetime


class MySubmitterContext(PortalModel):
    assigned: bool
    submitter_id: int | None = None
    submitter_name: str | None = None
    property_name: str | None = None
    property_place: str | None = None
    unit: str | None = None
    meter_name: str | None = None
    latest_reading: MyMeterReading | None = None
    tariff: list[ElectricitySlab] = Field(default_factory=list)


class MyMeterReadingCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    current_reading: Decimal = Field(ge=0, max_digits=18, decimal_places=6, allow_inf_nan=False)
    reading_date: date = Field(default_factory=date.today)


class MyMeterReadingSubmitResponse(PortalModel):
    reading: MyMeterReading
    high_usage_warning: str | None = None


class MyBill(PortalModel):
    rate_mode: Literal["manual", "slab"] | None = None
    breakdown: list[BreakdownItem] = Field(default_factory=list)
    meter_name: str | None = None
    id: int
    invoice_number: str
    submitter_name: str
    property_name: str | None
    unit: str | None
    billing_period_start: date
    billing_period_end: date
    issued_at: datetime
    due_date: date
    status: Literal["pending", "overdue", "paid"]
    previous_reading: Decimal | None = None
    current_reading: Decimal | None = None
    units: Decimal
    rate_per_unit: Decimal | None = None
    tariff_label: str
    total_amount: Decimal
    payment_method: Literal["upi", "card", "net_banking"] | None = None
    transaction_id: str | None = None
    paid_at: datetime | None = None


class MyBillListResponse(PortalModel):
    items: list[MyBill]
    total: int


class DemoPaymentRequest(BaseModel):
    """Only the method is client-selected; payment state is server-owned."""

    model_config = ConfigDict(extra="forbid")

    payment_method: Literal["upi", "card", "net_banking"]


class ConsumptionPoint(PortalModel):
    month: str
    label: str
    units: Decimal


class MyConsumption(PortalModel):
    current_month_units: Decimal
    previous_month_units: Decimal
    absolute_change: Decimal
    percentage_change: Decimal | None
    history: list[ConsumptionPoint]

File: frontend/src/types/portal.ts
import type { BreakdownItem, ElectricitySlab } from "@/types/electricity";

export type MyMeterReading = {
  id: number;
  meter_name: string;
  previous_reading: number;
  current_reading: number;
  units_used: number;
  reading_date: string;
  created_at: string;
};

export type MySubmitterContext = {
  assigned: boolean;
  submitter_id: number | null;
  submitter_name: string | null;
  property_name: string | null;
  property_place: string | null;
  unit: string | null;
  meter_name: string | null;
  latest_reading: MyMeterReading | null;
  tariff: ElectricitySlab[];
};

export type MyMeterReadingInput = {
  current_reading: number;
  reading_date: string;
};

export type MyMeterReadingSubmission = {
  reading: MyMeterReading;
  high_usage_warning: string | null;
};

export type MyBillStatus = "pending" | "overdue" | "paid";

export type MyBill = {
  rate_mode?: "manual" | "slab" | null;
  meter_name?: string | null;
  breakdown?: BreakdownItem[];
  id: number;
  invoice_number: string;
  submitter_name: string;
  property_name: string | null;
  unit: string | null;
  billing_period_start: string;
  billing_period_end: string;
  issued_at: string;
  due_date: string;
  status: MyBillStatus;
  previous_reading: number | null;
  current_reading: number | null;
  units: number;
  rate_per_unit: number | null;
  tariff_label: string;
  total_amount: number;
  payment_method?: "upi" | "card" | "net_banking" | null;
  transaction_id?: string | null;
  paid_at?: string | null;
};

export type MyBillListResponse = { items: MyBill[]; total: number };

export type ConsumptionPoint = { month: string; label: string; units: number };

export type MyConsumption = {
  current_month_units: number;
  previous_month_units: number;
  absolute_change: number;
  percentage_change: number | null;
  history: ConsumptionPoint[];
};
